crates/pop-cli/src/commands/up/parachain.rs

```diff
@@ -0,0 +1,415 @@
+					.default_input(DEFAULT_URL)
+					.interact()?;
+				Url::parse(&url)?
+			},
+		};
+
+		// Parse metadata from chain url.
+		let client = set_up_client(url.as_str()).await?;
+		let pallets = parse_chain_metadata(&client).map_err(|e| {
+			anyhow!(format!("Unable to fetch the chain metadata: {}", e.to_string()))
+		})?;
+		Ok(Chain { url, client, pallets })
+	}
+	// Resolves the parachain ID, reserving a new one if necessary.
+	async fn resolve_parachain_id(&self, chain: &Chain, cli: &mut impl Cli) -> Result<u32> {
+		match self.id {
+			Some(id) => Ok(id),
+			None => {
+				cli.info("Reserving a parachain ID")?;
+				reserve_para_id(chain, cli).await
+			},
+		}
+	}
+	// Resolves the genesis state and code files, generating them if necessary.
+	async fn resolve_genesis_files(
+		&self,
+		para_id: u32,
+		cli: &mut impl Cli,
+	) -> Result<(PathBuf, PathBuf)> {
+		match (self.genesis_state.clone(), self.genesis_code.clone()) {
+			(Some(state), Some(code)) => Ok((state, code)),
+			_ => {
+				cli.info("Generating the chain spec for your parachain")?;
+				generate_spec_files(para_id, self.path.clone(), cli).await
+			},
+		}
+	}
+}
+
+// Represents the configuration for deploying a chain.
+pub(crate) struct UpChain {
+	id: u32,
+	genesis_state: PathBuf,
+	genesis_code: PathBuf,
+	chain: Chain,
+}
+impl UpChain {
+	// Registers a parachain by submitting an extrinsic.
+	async fn register_parachain(&self, cli: &mut impl Cli) -> Result<()> {
+		cli.info("Registering a parachain ID")?;
+		let call_data = self.prepare_register_parachain_call_data()?;
+		submit_extrinsic_with_wallet(&self.chain.client, &self.chain.url, call_data, cli).await?;
+		Ok(())
+	}
+
+	// Prepares and returns the encoded call data for registering a parachain.
+	fn prepare_register_parachain_call_data(&self) -> Result<Vec<u8>> {
+		let UpChain { id, genesis_code, genesis_state, chain } = self;
+		let ex = find_dispatchable_by_name(
+			&chain.pallets,
+			Action::Register.pallet_name(),
+			Action::Register.function_name(),
+		)?;
+		let state = std::fs::read_to_string(genesis_state)
+			.map_err(|err| anyhow!("Failed to read genesis state file: {}", err.to_string()))?;
+		let code = std::fs::read_to_string(genesis_code)
+			.map_err(|err| anyhow!("Failed to read genesis state file: {}", err.to_string()))?;
+		let xt = construct_extrinsic(ex, vec![id.to_string(), state, code])?;
+		Ok(xt.encode_call_data(&chain.client.metadata())?)
+	}
+}
+
+// Reserves a parachain ID by submitting an extrinsic.
+async fn reserve_para_id(chain: &Chain, cli: &mut impl Cli) -> Result<u32> {
+	let call_data = prepare_reserve_parachain_call_data(chain)?;
+	let events = submit_extrinsic_with_wallet(&chain.client, &chain.url, call_data, cli)
+		.await
+		.map_err(|e| anyhow::anyhow!("Parachain ID reservation failed: {}", e))?;
+	let para_id = extract_para_id_from_event(&events).map_err(|_| {
+		anyhow::anyhow!("Unable to parse the event. Specify the parachain ID manually with --id.")
```

Comment:
```suggestion
		anyhow::anyhow!(r#"Unable to parse the event. Specify the parachain ID manually with "--id"."#)
```

I guess this improves the readability